core/domain/exp_domain.py

```diff
@@ -2514,12 +2515,1064 @@ def _convert_states_v50_dict_to_v51_dict(
+        before `HasElementXAtPositionY` otherwise the rule will never going to
+        match, this helper functions simply removes the `Equals` rules as it
+        will never going to match
+
+        Args:
+            answer_groups: List[state_domain.AnswerGroup]. The answer group.
+
+        Returns:
+            List[state_domain.AnswerGroup]. The modified answer group.
+        """
+        ele_x_at_y_rules = []
+        rules_to_remove = []
+        for answer_group in answer_groups:
+            for rule_spec in answer_group['rule_specs']:
+                if rule_spec['rule_type'] == 'HasElementXAtPositionY':
+                    element = rule_spec['inputs']['x']
+                    position = rule_spec['inputs']['y']
+                    ele_x_at_y_rules.append(
+                        {'element': element, 'position': position}
+                    )
+
+                if rule_spec['rule_type'] == 'IsEqualToOrdering':
+                    if len(rule_spec['inputs']['x']) <= 0:
+                        rules_to_remove.append(rule_spec)
+                        continue
+                    for ele in ele_x_at_y_rules:
+                        ele_position = ele['position']
+                        ele_element = ele['element']
+                        rule_choice = rule_spec['inputs']['x'][
+                            ele_position - 1]
+
+                        if len(rule_choice) > 1:
+                            for choice in rule_choice:
+                                if choice == ele_element:
+                                    rules_to_remove.append(rule_spec)
+                        else:
+                            if len(rule_choice) <= 0:
+                                rules_to_remove.append(rule_spec)
+                                continue
+                            if rule_choice[0] == ele_element:
+                                rules_to_remove.append(rule_spec)
+
+        for rule_to_remove in rules_to_remove:
+            for answer_group in answer_groups:
+                for rule_spec in answer_group['rule_specs']:
+                    if rule_to_remove == rule_spec:
+                        answer_group['rule_specs'].remove(rule_to_remove)
+
+        return answer_groups
+
+    @classmethod
+    def _equals_should_come_before_misplace_by_one_rule(
+        cls, answer_groups: List[state_domain.AnswerGroup]
+    ) -> List[state_domain.AnswerGroup]:
+        """Inside DragAndDrop interaction the `Equals` rule should always come
+        before `IsEqualToOrderingWithOneItemAtIncorrectPosition` otherwise the
+        rule will never going to match, this helper functions simply removes
+        the `Equals` rules as it will never going to match
+
+        Args:
+            answer_groups: List[state_domain.AnswerGroup]. The answer group.
+
+        Returns:
+            List[state_domain.AnswerGroup]. The modified answer group.
+        """
+        equal_ordering_one_at_incorec_posn = []
```

Comment:
To follow the convention you started above with `ele_x_at_y_rules`, can we do:
```suggestion
        off_by_one_rules = []
```